### PR TITLE
Initial Workflow Rework

### PR DESCRIPTION
**Pull request by epikichi**
# Description

This PR introduces a rework to our Github Actions workflows, as shown on our internal RFC. The proper Github environments needs to be set before merging for these workflows to function properly.

### Reusable workflows
- Pull Request CI
- DevNet Deployment 
- TestNet Deployment
- [Loadbot ](https://github.com/maticnetwork/go-loadbot) Performance Testing 
- [Pandora's Box](https://mirror.cnhub.dev/Trapesys/pandoras-box) Performance Testing
- GoSec Scan

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      GO111MODULE: on  # yamllint disable-line rule:truthy
+    steps:
+      - name: Checkout Source
+        uses: actions/checkout@v2
+      - name: Run Gosec Security Scanner
+        uses: securego/gosec@master
+        with:
+          args: ./...
+
+  notify-deployer:
```
Comment (ZeljkoBenovic) [inline .github/workflows/ci.yml]: We can delete the whole `notify-deployer` section as it is legacy code.

File: .github/workflows/deploy.testnet.yml [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    needs: deploy_testnet
+    secrets:
+      SLACK_PERFORMANCE_WEBHOOK_URL: ${{ secrets.SLACK_PERFORMANCE_WEBHOOK_URL }}
+      PANDORAS_TARGET: ${{ secrets.PANDORAS_TARGET }}
+      PANDORAS_MNEMONIC: ${{ secrets.PANDORAS_MNEMONIC }}
+    with:
+      runner: testnet
+      environment: testnet
+      transaction_batch: '20'
+      transaction_count: '2000'
```
Comment (ZeljkoBenovic) [inline .github/workflows/deploy.testnet.yml]: Maybe we should increase the number of transactions to at least 10k , as 2k is relatively small transaction volume which won't properly stress test the network?

**Review by ZeljkoBenovic — APPROVED** [review]
Great job 🥇 
Left some minor comments.

File: .github/workflows/deploy.testnet.yml (hunk 2 above)
Comment (epikichi) [reply .github/workflows/deploy.testnet.yml]: Tx Count updated to `10000` for both Pandora's Box and Loadbot in https://github.com/0xPolygon/polygon-edge/pull/765/commits/d3f9c33ccaaac75a013cbe3809c55518b1a61455

File: .github/workflows/deploy.devnet.yml [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                    "type": "mrkdwn",
+                    "text": "Devnet Deployment Status: ${{ job.status }}\n<${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}|Job Status>"
+                  }
+                }
+              ]
+            }
+
+      - name: Publish Binary package to S3
+        run: |
+          aws s3 cp ./polygon-edge.tar.gz s3://${{ secrets.POLYGON_EDGE_ARTIFACT_BUCKET }} --metadata "{\"commit_sha\":\"${GITHUB_SHA}\"}"
```
Comment (vcastellm) [inline .github/workflows/deploy.devnet.yml]: Are we going to maintain historic versions of the package?

File: .github/workflows/deploy.devnet.yml [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+      - name: Deploy DevNet
+        continue-on-error: true
+        env:
+          VALIDATOR_ASGS: ${{ secrets.VALIDATOR_ASGS }}
+        run: |
+          # Parallel deploy all nodes, relying on ALB Health Checks
```
Comment (vcastellm) [inline .github/workflows/deploy.devnet.yml]: Are we maintaining the data/volumes between deployments?

**Comment by epikichi** [comment]
Expected to resolve conflicts from https://github.com/0xPolygon/polygon-edge/pull/776

File: .github/workflows/deploy.devnet.yml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                    "type": "mrkdwn",
+                    "text": "Devnet Deployment Status: ${{ job.status }}\n<${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}|Job Status>"
+                  }
+                }
+              ]
+            }
+
+      - name: Publish Binary package to S3
+        run: |
+          aws s3 cp ./polygon-edge.tar.gz s3://${{ secrets.POLYGON_EDGE_ARTIFACT_BUCKET }} --metadata "{\"commit_sha\":\"${GITHUB_SHA}\"}"
```
Comment (vcastellm) [inline .github/workflows/deploy.devnet.yml:79]: Are we going to maintain historical versions of the package?

File: .github/workflows/deploy.testnet.yml [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+      - name: Publish Binary package to S3
+        run: |
+          aws s3 cp ./polygon-edge.tar.gz s3://${{ secrets.POLYGON_EDGE_ARTIFACT_BUCKET }} --metadata "{\"commit_sha\":\"${GITHUB_SHA}\"}"
+      - name: Deploy to Testnet
+        continue-on-error: true
+        env:
+          VALIDATOR_ASGS: ${{ secrets.VALIDATOR_ASGS }}
+        run: |
+          echo "Deploying new TestNet..."
```
Comment (vcastellm) [inline .github/workflows/deploy.testnet.yml:83]: Are we going to reset the data between deployments in testnets?

File: .github/workflows/deploy_edgenet.yaml [hunk 7]
```diff
@@ -22,8 +23,8 @@ jobs:
           AWS_ACCESS_KEY_ID: ${{ secrets.EDGENET_AWS_ACCESS_KEY_ID }}
           AWS_SECRET_ACCESS_KEY: ${{ secrets.EDGENET_AWS_SECRET_ACCESS_KEY }}
         run: >
-          /usr/local/bin/aws-commander 
-          -instances i-039f7c0b3328a00f8,i-035b9f2d78cfb8ea9,i-00a6c7cb3a213f21f,i-03ac2f42ddcba6120 
-          -mode ansible 
-          -playbook .github/workflows/ansible/update_edgenet.yaml 
+          /usr/local/bin/aws-commander
+          -instances i-039f7c0b3328a00f8,i-035b9f2d78cfb8ea9,i-00a6c7cb3a213f21f,i-03ac2f42ddcba6120
```
Comment (vcastellm) [inline .github/workflows/deploy_edgenet.yaml:27]: Could we implement a dynamic inventory based on tags here?

File: .github/workflows/deploy.testnet.yml (hunk 6 above)
Comment (epikichi) [reply .github/workflows/deploy.testnet.yml:83]: Not for Testnet. The Testnet deployment currently initiates instance refreshes on each ASG **sequentially**. On init via user-data, the new EC2s attach their respective EBS with the chain data persistent.

DevNet deployment however runs an SSM document that stops the service and cleans up the data directory prior to initiating instance refresh **asynchronously**

An alternative Testnet deployment is running SSM to handle the deployment scope on the instance without a re-cycle.

File: .github/workflows/deploy_edgenet.yaml (hunk 7 above)
Comment (epikichi) [reply .github/workflows/deploy_edgenet.yaml:27]: Yes we can. Edgenet deployment wasn't in this PR scope but we discussed about using tags or resource groups initially.

I can implement this logic for the `VALIDATOR_ASG` secret variable though for the proposed `devnet` and `testnet` deployemnts. :pray: Thanks for the suggestion!

File: .github/workflows/deploy.devnet.yml (hunk 5 above)
Comment (epikichi) [reply .github/workflows/deploy.devnet.yml:79]: [This line ](https://github.com/0xPolygon/polygon-edge/pull/765/files/30de509c23441317ff92cd7c7db9f290ee8949ea#diff-5b99851af7508b7aa954e62ebb2d8ae12f67239464e467fff03fe1a7994056eaR91-R92) sends an SSM command  that stops the polygon-edge service and removes the node's chain data.

 TestNet deployment does not run the cleanup command.

Comment (epikichi) [reply .github/workflows/deploy.devnet.yml:79]: For `devnet` and `testnet`, we can use S3 versions with the `commit_sha` metadata. I am open to discuss on a more formalized approach for artifacting here.

File: .github/workflows/deploy.testnet.yml (hunk 6 above)
Comment (vcastellm) [reply .github/workflows/deploy.testnet.yml:83]: LGTM

File: .github/workflows/deploy.devnet.yml (hunk 5 above)
Comment (vcastellm) [reply .github/workflows/deploy.devnet.yml:79]: S3 versioning should be 👌  at this point.

**Review by vcastellm — APPROVED** [review]
LGTM

**Review by Kourin1996 — APPROVED** [review]

**Review by zivkovicmilos — APPROVED** [review]
Looks great 💯

**Pull request merged**